### ISSUE TITLE
[InferenceRunner] input stopped before reaching its __len__()! You might need to check your input implementation.

### ISSUE THREAD

**Issue by ideaRunner** [tensorpack/tensorpack#949]
Hi，

I am writing codes based on example [mnist-keras.py](https://github.com/tensorpack/tensorpack/blob/master/examples/keras/mnist-keras.py) .

Here I got an Error:
```
[1027 10:23:08 @base.py:233] Initializing the session ...
[1027 10:23:08 @base.py:240] Graph Finalized.
[1027 10:23:08 @concurrency.py:37] Starting EnqueueThread QueueInput/input_queue ...
[1027 10:23:08 @inference_runner.py:101] [InferenceRunner] Will eval 38 iterations
[1027 10:23:08 @base.py:272] Start Epoch 1 ...
100%|############################################################################################|341/341[01:51<00:00, 0.14it/s]
[1027 10:24:59 @base.py:282] Epoch 1 (global_step 341) finished, time:1 minute 51 seconds.
[1027 10:25:01 @saver.py:77] Model saved to train_log/train_with_tensorpack/model-341.
  0%|                                                                                                       |0/38[00:00<?,?it/s]
[1027 10:25:01 @inference_runner.py:54] ERR [InferenceRunner] input stopped before reaching its __len__()! You might need to check your input implementation.
```

As I see, training epoch is finished safely, my problem comes from the `InferenceRunner`.

```
            InferenceRunner(
                dataset_test,
                ScalarStats(['cross_entropy_loss']))
```
This is my `InferenceRunner` used in `TrainConfig`.

and my `dataset_train` and `dataset_test` comes from the 2 lmdb files I write before.

Since no error with training epoch and 2 lmdb files generated by the same codes by different image files, I roughly think that the problem is not comes from my lmdb files writing.

I read the Error information, it seems that they cannot read all data from `BatchData`, I guess the problem is that I didn't add a parameter  `remainder=True` when generate `BatchData` of `dataset_test`.

I changed to add it, but same Error comes again. It looks like the `remainder=True` parameter of  `BatchData` is not the problem that it cannot reach its __len__().

I don't know how to handle it with the only information shown.

My enviroment:

```
Python 2.7.5
tf.git_version :  v1.10.0-0-g656e7a2b34
tensorflow version : 1.10.0
tensorpack version : 0.8.9
GPU: 1 x Tesla P100
```

**Comment by ppwwyyxx** [comment]
Please check whether `dataset_test.__len__()` is correct and see if `for cnt, data in enumerate(dataset_test): print(cnt)` can produce this many of data.

**Comment by ideaRunner** [comment]
For my codes, it returns an error:

**Comment by ideaRunner** [comment]
This it the code that how I generate lmdb file.

I read raw jpeg data so that the lmdb file will not be too much larger than origin images' size.

I designed a object, designed __len__ and __iter__, it looks no problem.

```python
def generate_multi_label_lmdb(save_lmdb_path, image_to_labels):
    """
    Generate lmdb file contain raw jpg format images, and a numpy array of multi-labels.
    :param save_lmdb_path: path to save lmdb file
    :param image_to_labels: a dict, key is image path, value is string labels, eg. "1 0 1 0 0 1"
    :return:
    """

    from tensorpack.dataflow import PrefetchDataZMQ, LMDBSerializer, DataFlow

    class ImageLabel(DataFlow):
        def __init__(self, dict_image_labels):
            self.dict = dict_image_labels

        def __len__(self):
            return len(self.dict)

        def __iter__(self):
            for img_path in self.dict:
                # img = np.array(Image.open(img_path))
                image_raw = open(img_path).read()
                labels = str(image_to_labels[img_path])
                labels_list = labels.split(" ")
                labels_list = [float(i) for i in labels_list]
                labels_array = np.array(labels_list)
                yield [image_raw, labels_array]

    ds0 = ImageLabel(image_to_labels)
    ds1 = PrefetchDataZMQ(ds0, nr_proc=1)
    LMDBSerializer.save(ds1, save_lmdb_path)
```

The codes below is for decode lmdb file.

```
def decode_multi_label_lmdb(load_lmdb_path, batchsize=48,  remainder=False):
    """
    Decode lmdb file and returan a BatchData
    :param load_lmdb_path: lmdb file path
    :param batchsize: batchsize of BatchData
    :param remainder: according BatchData reminder
    :return: BatchData [images_array, labels_array]
    """
    import cv2
    from tensorpack.dataflow import BatchData, LMDBSerializer, MapDataComponent, LocallyShuffleData, AugmentImageComponent
    ds = LMDBSerializer.load(load_lmdb_path, shuffle=False)
    ds = LocallyShuffleData(ds, 50000)
    ds = MapDataComponent(ds, lambda x: np.asarray(bytearray(x), dtype="uint8"), 0)
    ds = MapDataComponent(ds, lambda x: cv2.imdecode(x, cv2.IMREAD_COLOR), 0)
    ds = MapDataComponent(ds, lambda x: x/float(255), 0)
    # ds = AugmentImageComponent(ds, lots_of_augmentors)
    ds = BatchData(ds, batchsize, remainder=remainder)
    return ds
```

**Comment by ppwwyyxx** [comment]
The issue is that `LocallyShuffleData` for now does not handle the case when the buffer size (50000) is even larger than the whole dataflow -- this case does not make sense at all. I'll add a warning for this.

Also, using LocallyShuffleData for inference is wrong at the first place because it randomly mixes datapoints.

**Comment by ideaRunner** [comment]
Hi, 

It is indeed the problem of `LocallyShuffleData `. Now it works. Thank you.

I followed this [document **Sequential Read**](https://tensorpack.readthedocs.io/tutorial/efficient-dataflow.html#sequential-read). It would be better if the Error mention about it or there has a warning.

emmm, there is another GPU utilization question refer to **Sequential Read**. I am not sure if I need to raise another issue. Could I write it here?

When I training with the lmdb decode codes I written before(commented out `LocallyShuffleData`, fixed problem), The GPU utilization is almost 30% all training epoch time and about 10% validation epoch.

It seems that `ds = MapDataComponent(ds, lambda x: cv2.imdecode(x, cv2.IMREAD_COLOR), 0)` cost too much CPU time.

At before, I use `tf.image.decode_jpeg()` function to decode raw jpeg data in `**tfrecord**` file, it looks well and my tensorflow codes can achieve almost 90% utilization.

But I tried here, seems that `lambda` works not well with it.

As I am not well understood tensorpack, is there any example that could use gpu to decode jpeg raw data? Otherwise, how can I achieve better GPU utilization mentioned in this [document](https://tensorpack.readthedocs.io/tutorial/input-source.html)?

**Comment by ppwwyyxx** [comment]
https://tensorpack.readthedocs.io/tutorial/performance-tuning.html

**Comment by ideaRunner** [comment]
Hi， 

The bottleneck is the data.

```
QueueInput/queue_size: 2.0649e-38
```

I know that [Efficitent DataFlow/Sequential Read](https://tensorpack.readthedocs.io/tutorial/efficient-dataflow.html#sequential-read) should solve that problem, as the codes I wrote above, but it seems didn't do what I thought.

I tried to write the whole numpy array in **lmdb generating** part to reduce the time of **lmdb decoding** cost. Besides I got a large lmdb file, however the queus_size is similar (near to zero), and GPU utilization is still low. training 30% val 10%. 

Do you have any suggestion?

**Comment by ppwwyyxx** [comment]
Yes the bottleneck is the data but https://tensorpack.readthedocs.io/tutorial/performance-tuning.html#investigate-dataflow has more information on figuring out __which part of data__ is the bottleneck. Before doing that it does not make sense to do any optimization.
